docs/source/app_developers_guide/about_event_subscriptions.rst

```diff
@@ -0,0 +1,137 @@
+**************************
+About Events Subscriptions
+**************************
+
+An event subscription consists of an event type (the name of the event) and an
+optional list of filters for this event type.
+
+An event is “in a subscription” if the event’s ``event_type`` field matches the
+subscription’s ``event_type`` field, and the event matches any filters that are
+in the event subscription. If multiple filters are included in a subscription,
+only events that pass *all* filters will be received.
+
+
+Event Filters
+=============
+
+An event filter specifies which event attributes to match. A filter has the
+following contents:
+
+  * A string describing the filter (the filter name)
+
+  * The filter type (simple or regular expression) and match type (any or all)
+
+An event filter consists of a key and a "match string". Filters apply only to
+the event attributes that have the same key as the filter. The filter's match
+string is compared against the attribute's value according to the type of
+event filter.
+
+The event filter type can either be ``SIMPLE`` (an exact match to the specified
+string) or ``REGEX`` (a match using a regular expression). In addition, the
+filter type can specify ``ANY`` (match one or more) or ``ALL`` (must match all).
+The following event filters are supported:
+
+* ``SIMPLE_ANY``:
+   The filter's "match string" must match at least one of the
+   event's attribute values. For example, this filter type with the match string
+   "``abc``" would succeed for an event with the following attributes, because
+   it matches the attribute value ``abc``.
+
+   .. code-block:: none
+
+      Attribute(key="address",value="abc")
+      Attribute(key="address",value="def")
+
+* ``SIMPLE_ALL``:
+   The filter's "match string" must match all of the event's
+   attribute values. This filter type with the match string "``abc``"
+   would fail with the previous example, because it does not match all the
+   attribute values.
+
+* ``REGEX_ANY``:
+   The filter's regular expression must evaluate to a match for
+   one or more of the event's attribute values. For example, this filter type
+   with the match string "``[ad][be][cf]``" would succeed for an event with the
```

Comment:
typo? "``[ad][be]*[cf]``"

Reply:
Nope.  I was trying to show a REGEX filter that would match some but not all of the following attribute values.  The * makes it match them all, which is not what I want.

Given my info, is there still a typo in my regex?  I'm a bit of a regex newbie.

Reply:
(Never mind - I deleted my previous comment because you're right and I'm wrong.)

Reply:
The end of the following sentence is where the error is.  It should say '...because it matches the attributes "abc" and "def".'

Reply:
There is still a typo in the regex. `[]` is "Used to indicate a set of characters" that you are accepting. `*` is used to indicate you want to accept 0 or more instances of some set of characters.

What you have says "accept any 3 character string where the first character is a or d, the second character is b or e, and the third character is c or f". What I think you want is "accept any string where the first character is a or d, the last character is c or f, and there are 0 or more characters in the middle that are either b or e". This is `[ad][be]*[cf]`.

Here is a little script that compares the two:
```
import re

def test_str(r, s):
  if r.match(s) is None:
    print("%s didn't match" % s)
  else:
    print("%s matched" % s)

def test_all(match_string):
  print("testing %s" % match_string)
  tests = [
    "abc",
    "abbbc",
    "def",
    "ac",
  ]

  r = re.compile(match_string)
  for t in tests:
    test_str(r, t)

test_all("[ad][be][cf]")
test_all("[ad][be]*[cf]")
```

```
~ » python3 testre.py
testing [ad][be][cf]
abc matched
abbbc didn't match
def matched
ac didn't match
testing [ad][be]*[cf]
abc matched
abbbc matched
def matched
ac matched
```

Reply:
Gah, I misunderstood what you were trying to say here. You were using the right regex for what you were trying to do.

Reply:
I've simplified the first regex expression to make my intent for the examples clearer.  I changed it to "ab." so that it matches only "abc" but not "abbbc" or "def".  Then, in the REGEX_ALL blurb, I said that "ab." fails but "[ad][be]*[cf]" passes.